### PR TITLE
Tickets/dm 6813

### PR DESCRIPTION
**Pull request by jgates108**

### CODE REVIEW

File: core/modules/wcontrol/Foreman.h [hunk 1]
```diff
@@ -28,6 +28,7 @@
 #include <atomic>
 #include <memory>
 
+#include "../wpublish/QueryChunkStatistics.h"
```
Comment (andy-slac) [inline core/modules/wcontrol/Foreman.h]: Do you need `../` here? Move it to `// Qserv headers` block?

File: core/modules/wpublish/QueryChunkStatistics.cc [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+// Class header
+#include "lsst/log/Log.h"
+#include "QueryChunkStatistics.h"
+
+// Qserv headers
+
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.wsched.QueryStatistics");
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.cc]: `wpublish` instead of `wsched`?

File: core/modules/wpublish/QueryChunkStatistics.cc [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "lsst/log/Log.h"
+#include "QueryChunkStatistics.h"
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.cc]: I believe you are still supposed to have package name in the include even if it is the same package (`wpublish/QueryChunkStatistics.h`)

File: core/modules/wpublish/QueryChunkStatistics.cc [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "lsst/log/Log.h"
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.cc]: This include should be moved below

File: core/modules/wpublish/QueryChunkStatistics.cc [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    }
+    QueryStatistics::Ptr stats = res.first->second;
+    guardStats.unlock();
+    stats->addTask(task);
+}
+
+
+/// Update statistics for the Task that was just queued.
+void QueryChunkStatistics::queuedTask(wbase::Task::Ptr const& task) {
+    std::chrono::system_clock::time_point now = std::chrono::system_clock::now();
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.cc]: `auto` saves reviewer's time! :smile_cat:

File: core/modules/wpublish/QueryChunkStatistics.cc [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        auto end = _deadList.end();
+        for (;iter != end; ++iter) {
+            if ((*iter)->isDead(_deadAfter, now)) {
+                dList.push_back(*iter);
+                iter = _deadList.erase(iter);
+            }
+        }
+    }
+
+    for (auto const& dead:dList) {
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.cc]: space after colon will make it less confusing.

File: core/modules/wpublish/QueryChunkStatistics.h [hunk 7]
```diff
@@ -0,0 +1,170 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2013-2016 LSST Corporation.
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.h]: 2016? Or was it moved from somewhere else?

File: core/modules/wpublish/QueryChunkStatistics.h [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_QSERV_WSCHED_QUERYCHUNKSTATISTICS_H
+#define LSST_QSERV_WSCHED_QUERYCHUNKSTATISTICS_H
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.h]: _WPUBLISH_?

File: core/modules/wpublish/QueryChunkStatistics.h [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+class QueryChunkStatistics;
+
+
+/// Statistics for a single user query.
+class QueryStatistics {
+public:
+    using Ptr = std::shared_ptr<QueryStatistics>;
+
+    explicit QueryStatistics(QueryId const& queryId) : _queryId{_queryId} {}
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.h]: ` _queryId{queryId}` (one extraneous underscore)

File: core/modules/wpublish/QueryChunkStatistics.h [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+    void addTask(wbase::Task::Ptr const& task);
+
+    bool isDead(std::chrono::seconds deadTime, std::chrono::system_clock::time_point now);
+
+    friend class QueryChunkStatistics;
+
+private:
+    bool _isMostlyDead();
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.h]: const? (or mostly const? :skull:)

File: core/modules/wpublish/QueryChunkStatistics.h [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+    double _totalCompletionTime{0.0};
+
+    std::map<int, wbase::Task::Ptr> _taskMap;
+};
+
+/// Statistics for a table in a chunk. Statistics are base on the slowest table in a query,
+/// so this most likely includes values for queries on _scanTableName and queries that join
+/// against _scanTableName. Source is slower than Object, so joins with Source and Object will
+/// have there statistics logged with Source.
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.h]: their

File: core/modules/wpublish/QueryChunkStatistics.h [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+    void addTaskFinished(double duration);
+
+private:
+    std::mutex _mtx;
+    int const _chunkId;
+    std::string const _scanTableName;
+
+    std::uint64_t _tasksCompleted{0}; ///< Number of Tasks that have completed on this chunk/table.
+    std::uint16_t _tasksBooted{0}; ///< Number of Tasks that have been booted for taking too long.
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.h]: This will never wrap?

File: core/modules/wpublish/QueryChunkStatistics.h [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+
+/// Statistics for one chunk.
+class ChunkStats {
+public:
+    using Ptr = std::shared_ptr<ChunkStats>;
+
+    ChunkStats(int chunkId) : _chunkId{chunkId} {}
+
+    ChunkStatsTable::Ptr add(std::string const& scanTableName, double duration);
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.h]: I'd make `duration` dimensional (like milliseconds) or at least change name to have _sec or _minutes suffix and comment it.

File: core/modules/wpublish/QueryChunkStatistics.h [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    ChunkStatsTable::Ptr getStats(std::string const& scanTableName) const;
+private:
+    int const _chunkId;
+    mutable std::mutex _tStatsMtx; ///< protects _tableStats;
```
Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.h]: why not all of your mutexes mutable?

File: core/modules/wsched/BlendScheduler.h [hunk 15]
```diff
@@ -26,7 +26,7 @@
 // System headers
 #include <map>
 
-// Qserv headers
```
Comment (andy-slac) [inline core/modules/wsched/BlendScheduler.h]: Why remove it?

File: core/modules/wsched/BlendScheduler.h [hunk 16]
```diff
@@ -26,7 +26,7 @@
...
+#include "../wpublish/QueryChunkStatistics.h"
```
Comment (andy-slac) [inline core/modules/wsched/BlendScheduler.h]: `../`?

File: core/modules/wsched/testSchedulers.cc [hunk 17]
```diff
@@ -30,6 +30,7 @@
 // Third-party headers
 
 // LSST headers
+#include "../wpublish/QueryChunkStatistics.h"
```
Comment (andy-slac) [inline core/modules/wsched/testSchedulers.cc]: I see a pattern here :smile:

Comment (jgates108) [inline core/modules/wsched/testSchedulers.cc]: I did a file move with eclipse. I didn't notice the additions.

File: core/modules/wpublish/QueryChunkStatistics.h (hunk 12 above)
Comment (jgates108) [inline core/modules/wpublish/QueryChunkStatistics.h]: That was supposed to be a unit64_t, thanks.

File: core/modules/wpublish/QueryChunkStatistics.h (hunk 9 above)
Comment (jgates108) [inline core/modules/wpublish/QueryChunkStatistics.h]: No warning for that :frowning:

Comment (andy-slac) [inline core/modules/wpublish/QueryChunkStatistics.h]: Hmm, interesting, I'd at least expect one for unused parameter. We probably have suppressed that sort of warnings for some reason.

File: core/modules/wpublish/QueryChunkStatistics.h (hunk 14 above)
Comment (jgates108) [inline core/modules/wpublish/QueryChunkStatistics.h]: I really like to think about something before declaring it mutable. On the other hand, I probably should be declaring more things const.

File: core/modules/wsched/testSchedulers.cc (hunk 17 above)
Comment (jgates108) [inline core/modules/wsched/testSchedulers.cc]: Or the other liberties it took with the affected includes...

**Pull request merged**